crates/oxc_linter/src/rules/react/require_render_return.rs

```diff
@@ -0,0 +1,243 @@
+    ///
+    /// ### Example
+    /// ```javascript
+    /// var Hello = createReactClass({
+    ///   render() {
+    ///     <div>Hello</div>;
+    ///   }
+    /// });
+    ///
+    /// class Hello extends React.Component {
+    ///   render() {
+    ///     <div>Hello</div>;
+    ///   }
+    /// }
+    /// ```
+    RequireRenderReturn,
+    correctness
+);
+
+impl Rule for RequireRenderReturn {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if !is_es5_component(node) && !is_es6_component(node) {
+            return;
+        }
+
+        match node.kind() {
+            AstKind::Class(cl) => {
+                if let Some((fn_body, is_arrow_function_expression)) =
+                    cl.body.body.iter().find_map(|ce| match ce {
+                        ClassElement::MethodDefinition(md)
+                            if md.key.is_specific_static_name("render") =>
+                        {
+                            md.value.body.as_ref().map(|v| (v, false))
+                        }
+                        ClassElement::PropertyDefinition(pd)
+                            if pd.key.is_specific_static_name("render") =>
+                        {
+                            if let Some(Expression::ArrowExpression(ref ae)) = pd.value {
+                                Some((&ae.body, ae.expression))
+                            } else {
+                                None
+                            }
+                        }
+                        _ => None,
+                    })
+                {
+                    if is_arrow_function_expression || has_return_in_fn_body(fn_body) {
+                        return;
+                    }
+
+                    ctx.diagnostic(RequireRenderReturnDiagnostic(fn_body.span));
+                }
+            }
+            AstKind::CallExpression(ce) => {
+                if let Some(Argument::Expression(Expression::ObjectExpression(obj_expr))) =
+                    ce.arguments.first()
+                {
+                    if let Some(fn_body) = obj_expr
+                        .properties
+                        .iter()
+                        .filter_map(|prop| match prop {
+                            ObjectPropertyKind::ObjectProperty(prop)
+                                if prop.key.is_specific_static_name("render") =>
+                            {
+                                if let Expression::FunctionExpression(ae) = &prop.value {
+                                    ae.body.as_ref()
+                                } else {
+                                    None
+                                }
+                            }
+                            _ => None,
+                        })
+                        .find(|fn_body| !has_return_in_fn_body(fn_body))
+                    {
+                        ctx.diagnostic(RequireRenderReturnDiagnostic(fn_body.span));
+                    }
+                }
```

Comment:
am reviewing fairly quickly, so correct me if i am wrong. But, won't this match

```ts
foo({ render: render: function() {} })
```
?

This is what the plugin reports for me (js not oxc)

<img width="809" alt="Screenshot 2024-01-09 at 22 43 56" src="https://github.com/oxc-project/oxc/assets/18101008/e9ec24ab-c576-412a-9147-c88536ec524c">


can we add a passing test case for the example above (unless i'm wrong lol)

Reply:
lol i was wrong - should've read the code more carefully.

above:
```rs
if !is_es5_component(node) && !is_es6_component(node) {
    return;
}
```

`react.rs`
```rs
const PRAGMA: &str = "React";
const CREATE_CLASS: &str = "createReactClass";

pub fn is_es5_component(node: &AstNode) -> bool {
    let AstKind::CallExpression(call_expr) = node.kind() else { return false };

    if let Expression::MemberExpression(member_expr) = &call_expr.callee {
        if let Expression::Identifier(ident) = member_expr.object() {
            return ident.name == PRAGMA
                && member_expr.static_property_name() == Some(CREATE_CLASS);
        }
    }

    if let Some(ident_reference) = call_expr.callee.get_identifier_reference() {
        return ident_reference.name == CREATE_CLASS;
    }

    false
}
```